### ISSUE TITLE
OnDestroy SteamAPI.Shutdown() is NOT working

### ISSUE THREAD

**Issue by flashmandv** [rlabrecque/Steamworks.NET#187]
Using the latest Unity SDK 10.0 (and also with the prev. version) I noticed that when I exit play mode from Unity - the SteamAPI is still active. In steam I still see my game as Running and the only way to stop it is to close the Unity itself.

This is very annoying because when I make a build and release it, I have to exit Unity every time so that Steam can see my game shutdown and update it later.

I debugged and saw that OnDestroy is called and SteamAPI.Shutdown() is called, but it is not working. 
Please check this out

**Comment by Alloc86** [comment]
That's not on Riley's end, that's how Steam works. If a process registers with Steam as a game Steam considers that game to be running until that process (and any child process that the process might have spawned afterwards) is killed.

**Comment by flashmandv** [comment]
So this means the Steam API method SteamAPI_Shutdown() is pretty much useless ?

**Comment by Alloc86** [comment]
Never seen any real purpose for it but not sure if there's anything else it helps with.

**Comment by rlabrecque** [comment]
It's not actually useless for Unity because what it does is cleans up Steams internal memory. If you call SteamAPI.Init many times over the lifetime of the process (say each time you play in the editor) without calling Shutdown that's just creating garbage that never gets free'd on Steam's side. It may or may not clear all of that when your process shuts down. (It will definitely automatically clear the memory from the most recent call to Init automatically, but might not clear the older ones.)

**Comment by flashmandv** [comment]
I just saw that when my game is started from Steam and then closed, it still appears as Running. Unity Editor have nothing to do with it. My other games are OK. But my own game continues to appear as running.
Do you have any ideas why can I fix this?

**Comment by rlabrecque** [comment]
Might be #151 ?

**Comment by Chaser324** [comment]
@flashmandv I've had that issue a few times as well, but I think it is still due to something lingering after the Unity Editor has been closed (maybe a library that is just sticking around in memory and refuses to close - I don't really know). I've never had any players actually complain about it, and rebooting my PC generally fixes it.

**Comment by flashmandv** [comment]
Apparently there is some internal crash (either in Unity plugin or in native SDK). So every time the players close the game - it hangs.  Some people say it is Steam bug..but most games in Steam are fine..so it should be Unity plugin bug.

I have another Steam game that is made with Adobe AIR and the Steam native extension for Adobe AIR works fine. The game is closed normally and Steam registers it (comparing to the crashing exit of my Unity game)

**Comment by StefDevs** [comment]
@rlabrecque @Chaser324 I was going to open a similar issue but I'll chime in here. 

With some testing I've found that this issue only occurs for me if SteamVR is not running before I launch the game. Since the game is starting the SteamVR processes, they are registered as sub-processes to the game. Steam will not recognize the game as being shut down until SteamVR is stopped.

Is there any way decouple these processes? I've noticed that this occurs with both Rec Room and Space Pirate Traners as well, but SPT seems to have a working solution since the SteamVR processes are decoupled upon shut down and the game closes successfully.

**Comment by eyalfx** [comment]
same issue here... using Unity5.6 .
SteamAPI.Shutdown(); is getting called, but the app still shows as running.

**Comment by JashanChittesh** [comment]
I found this issue looking for https://github.com/rlabrecque/Steamworks.NET/issues/13 which has some really good explanations and troubleshooting tips for this (starting with @rlabrecque 's [comment on Jan 15](https://github.com/rlabrecque/Steamworks.NET/issues/13#issuecomment-70172801)).

In a nutshell: When running the game inside Unity, any process that was started by Unity will keep the game "running" in Steam. That includes Visual Studio, Rider or any IDE that Unity opens. It also includes processes like `adb.exe`, or, as I found out after doing some process dependency investigation, `VBCSCompiler.exe` and `conhost.exe`. Or, as @StefDevs said, SteamVR. And of course, Unity itself still running (exiting play mode obviously doesn't close the Unity process - and that's what Steam is looking at).

So this is not an issue with Steamworks.NET, not even an issue with your game but just an issue with how Steam shows games as "running". For that reason, unless you start new processes in your game that you don't properly close when closing the game (or something else is broken in your game), this won't be an issue when the game runs through Steam. It's very specific to running your game in the Unity editor.

The way I deal with this is making sure that my IDE and SteamVR are always running before even starting Unity. That way, when closing Unity, the game _usually_ stops "running" in Steam. Only sometimes, another process gets stuck and I need to look up this thread to remember which ones I need to kill ;-)

With the current Steam beta (that has the new library design), I'm also seeing Steam apparently shutting down with Unity quite frequently, which results in the game not being able to connect to Steam after I restart Unity.
